signature/cose/cose.go

```diff
@@ -15,34 +26,472 @@ func init() {
+	keySpec, err := signature.ExtractKeySpec(signingCert)
+	if err != nil {
+		return 0, err
+	}
+	return getSignatureAlgorithmFromKeySpec(keySpec)
+}
+
+func getSignatureAlgorithmFromKeySpec(keySpec signature.KeySpec) (cose.Algorithm, error) {
+	switch keySpec.Type {
+	case signature.KeyTypeRSA:
+		switch keySpec.Size {
+		case 2048:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmPS256], nil
+		case 3072:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmPS384], nil
+		case 4096:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmPS512], nil
+		default:
+			return 0, errors.New("key size not supported")
+		}
+	case signature.KeyTypeEC:
+		switch keySpec.Size {
+		case 256:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmES256], nil
+		case 384:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmES384], nil
+		case 512:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmES512], nil
+		default:
+			return 0, errors.New("key size not supported")
+		}
+	default:
+		return 0, errors.New("key type not supported")
+	}
+}
+
+func generateCoseProtectedHeaders(req *signature.SignRequest, protected cose.ProtectedHeader) error {
+	// crit, signingScheme, expiry, signingTime, authenticSigningTime
+	var crit []interface{}
+	crit = append(crit, headerKeySigningScheme)
+	protected[headerKeySigningScheme] = string(req.SigningScheme)
+	if !req.Expiry.IsZero() {
+		crit = append(crit, headerKeyExpiry)
+		protected[headerKeyExpiry] = uint(req.Expiry.Unix())
+	}
+	switch req.SigningScheme {
+	case signature.SigningSchemeX509:
+		protected[headerKeySigningTime] = uint(req.SigningTime.Unix())
+	case signature.SigningSchemeX509SigningAuthority:
+		crit = append(crit, headerKeyAuthenticSigningTime)
+		protected[headerKeyAuthenticSigningTime] = uint(req.SigningTime.Unix())
+	default:
+		return errors.New("SigningScheme: require notary.x509 or notary.x509.signingAuthority")
+	}
+	protected[cose.HeaderLabelCritical] = crit
+
+	// content type
+	protected[cose.HeaderLabelContentType] = req.Payload.ContentType
+
+	return nil
+}
+
+func sign(req *signature.SignRequest, signer signature.Signer, msgToSign *cose.Sign1Message, alg cose.Algorithm) ([]byte, []*x509.Certificate, error) {
+	var coseSigner cose.Signer
+	var err error
+	if localSigner, ok := signer.(signature.LocalSigner); ok {
+		switch key := localSigner.PrivateKey().(type) {
+		case *rsa.PrivateKey, *ecdsa.PrivateKey:
+			// Use go-cose's signer
+			coseSigner, err = cose.NewSigner(alg, key.(crypto.Signer))
+			if err != nil {
+				return nil, nil, err
+			}
+		default:
+			return nil, nil, errors.New("unsupported key type")
+		}
+	} else {
+		// Use External plugin's signer
+		coseSigner = pluginSigner{signer: signer}
+	}
+	err = msgToSign.Sign(rand.Reader, nil, coseSigner)
+	if err != nil {
+		return nil, nil, err
+	}
+	certs, err := signer.CertificateChain()
+	if err != nil {
+		return nil, nil, err
+	}
+	var certChain [][]byte
+	for i, c := range certs {
+		certChain[i] = c.Raw
```

Comment:
This line will always panic, `certChain` is not initialized, so indexed assignments panic.

Reply:
Code updated.